platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbCompactionCoordinator.java

```diff
@@ -0,0 +1,281 @@
+ * that there are no concurrent compactions for the same storage type. And finally it provides a way to stop all compactions
+ * and keep them disabled until they are explicitly enabled again.
+ * The compaction tasks are executed in a background thread pool.
+ * The number of threads in the pool is defined by {@link MerkleDbConfig#compactionThreads()} property.
+ *
+ */
+@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+class MerkleDbCompactionCoordinator {
+
+    /**
+     * Since {@code com.swirlds.platform.Browser} populates settings, and it is loaded before any
+     * application classes that might instantiate a data source, the {@link ConfigurationHolder}
+     * holder will have been configured by the time this static initializer runs.
+     */
+    private static final MerkleDbConfig config = ConfigurationHolder.getConfigData(MerkleDbConfig.class);
+
+    /**
+     * A thread pool to run compaction tasks.
+     */
+    private static final AtomicReference<ExecutorService> compactionExecutorServiceRef = new AtomicReference<>();
+
+    private static final Logger logger = LogManager.getLogger(MerkleDbCompactionCoordinator.class);
+
+    public static final String HASH_STORE_DISK_SUFFIX = "HashStoreDisk";
+    public static final String OBJECT_KEY_TO_PATH_SUFFIX = "ObjectKeyToPath";
+    public static final String PATH_TO_KEY_VALUE_SUFFIX = "PathToKeyValue";
+    private final AtomicBoolean compactionEnabled = new AtomicBoolean();
+    // we need a map of exactly three elements, one per storage
+    final ConcurrentMap<String, InterruptibleCompletableFuture<?>> compactionFuturesByName = new ConcurrentHashMap<>(3);
+    private final MerkleDbStatisticsUpdater statisticsUpdater;
+    private final Optional<CompactionTask> objectKeyToPathTaskOpt;
+    private final Optional<CompactionTask> hashesStoreDiskTaskOpt;
+    private final CompactionTask pathToKeyValueTask;
+
+    /**
+     * Creates a new instance of {@link MerkleDbCompactionCoordinator}.
+     * @param tableName the name of the table
+     * @param statisticsUpdater a statistics updater
+     * @param objectKeyToPath an object key to path store
+     * @param hashesStoreDisk a hash store
+     * @param pathToKeyValue a path to key-value store
+     */
+    public MerkleDbCompactionCoordinator(
+            @NonNull String tableName,
+            @NonNull MerkleDbStatisticsUpdater statisticsUpdater,
+            @Nullable Compactible objectKeyToPath,
+            @Nullable Compactible hashesStoreDisk,
+            @NonNull Compactible pathToKeyValue) {
+        this.statisticsUpdater = statisticsUpdater;
+        this.objectKeyToPathTaskOpt = Optional.ofNullable(objectKeyToPath)
+                .map(v -> new CompactionTask(
+                        tableName + OBJECT_KEY_TO_PATH_SUFFIX,
+                        v,
+                        statisticsUpdater::setLeafKeysStoreCompactionTimeMs,
+                        statisticsUpdater::setLeafKeysStoreCompactionSavedSpaceMb));
```

Comment:
This is using Optional just to use streaming and lambdas instead of a simple condition.  That is, in my opinion, a highly indesireable approach.
```suggestion
        this.objectKeyToPathTaskOpt = objectKeyToPath == null
                ? null
                : new CompactionTask(
                        tableName + OBJECT_KEY_TO_PATH_SUFFIX,
                        objectKeyToPath,
                        statisticsUpdater::setLeafKeysStoreCompactionTimeMs,
                        statisticsUpdater::setLeafKeysStoreCompactionSavedSpaceMb));
```
Again, `Optional` is not gaining anything here.  These are not method return types, and there is no semantic difference between "not set" and `null`, so there is no good reason to use `Optional`, in my opinion.

From the documentation for Optional (the API note):

> Optional is primarily intended for use as a method return type where there is a clear need to represent "no result," and where using null is likely to cause errors.

Reply:
I refactored the code. it's no longer using `Optional`-s